### PR TITLE
Add Azure-specific headers when uploading to blob storage

### PR DESCRIPTION
**Pull request by devictr**
# TL;DR
This PR fixes an issue causing `pyflyte run --remote`  to fail when using Azure. 

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [] Any pending items have an associated Issue

## Complete description
The PUT call to Azure Blob Storage returned an HTTP 400 Error: 
```
<?xml version="1.0" encoding="utf-8"?>
<Error>
    <Code>MissingRequiredHeader</Code>
    <Message>
        An HTTP header that's mandatory for this request is not specified.
        RequestId:ebd5dfbc-f01e-001d-5074-cb4f26000000
        Time:2023-08-10T10:24:58.6553767Z
    </Message>
    <HeaderName>x-ms-blob-type</HeaderName>
</Error>
```

This is a starting point that unblocks me, but there's a discussion to be had regarding where you would like to handle such cases. Some open questions:

1. Should this be done on the client? And if yes, is that the right place?
2. Or should the response that contains the signed URL also specify headers based on configuration options?

I have tested this locally by tweaking the class and trying to run a workflow

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3942

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: flytekit/remote/remote.py [hunk 1]
```diff
@@ -790,20 +790,23 @@ def upload_file(
...
                 data=content,
-                headers={"Content-Length": str(content_length), "Content-MD5": encoded_md5},
+                headers=headers,
                 verify=False
                 if self._config.platform.insecure_skip_verify is True
                 else self._config.platform.ca_cert_file_path,
             )
 
-            if rsp.status_code != requests.codes["OK"]:
+            if rsp.status_code not in (requests.codes["OK"], requests.codes["created"]):
```
Comment (eapolinario) [inline flytekit/remote/remote.py:809]: why do we have to check this value as well?

Comment (devictr) [reply flytekit/remote/remote.py:809]: Ah sorry, forgot to specify in my PR description. 
Azure returns an HTTP 201. So this was throwing an exception even though the request was successful

**Comment by devictr** [comment]
Maybe I should add a test on the entire `upload_file` function that also mocks the put call and checks that the headers are set. Might do that tomorrow

File: flytekit/remote/remote.py [hunk 2]
```diff
@@ -1925,3 +1928,9 @@ def launch_backfill(
             return remote_wf
 
         return self.execute(remote_wf, inputs={}, project=project, domain=domain, execution_name=execution_name)
+
+    @staticmethod
+    def get_extra_headers_for_protocol(native_url):
+        if native_url.startswith("abfs://"):
+            return {"x-ms-blob-type": "BlockBlob"}
```
Comment (pingsutw) [inline flytekit/remote/remote.py:1935]: In the case of blockblob type, the return code will become 200, won't it?
https://stackoverflow.com/questions/67015459/why-does-azure-blob-storage-return-201-but-not-upload

Comment (devictr) [reply flytekit/remote/remote.py:1935]: Seems like it returns a 201 https://learn.microsoft.com/en-us/rest/api/storageservices/put-blob?tabs=azure-ad#status-code

Comment (devictr) [reply flytekit/remote/remote.py:1935]: That matches what happened when I tested it

File: flytekit/remote/remote.py (hunk 1 above)
Comment (devictr) [reply flytekit/remote/remote.py:810]: Done

**Review by wild-endeavor — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
Thank you @devictr
